src/Simulation/Common/IQuantumProcessor.cs

```diff
@@ -391,6 +391,41 @@ public interface IQuantumProcessor
         /// <param name="qubits">Register to apply the exponent to.</param>
         void ControlledExpFrac(IQArray<Qubit> controls, IQArray<Pauli> paulis, long numerator, long power, IQArray<Qubit> qubits);
 
+        /// <summary>
+        /// Used as extensibility mechanism to add custom operations to IQuantumProcessor. 
+        /// </summary>
+        /// <remarks>
+        /// The custom operation may use only some of the parameters. Depending on the nature of the operation, all of them may be optional. 
+        /// A variety of parameter types are provided for flexibility.
+        /// </remarks>
+        /// <param name="operationId">The id of the custom operation. It is up to the custom operation implementor to define the meaning.</param>
```

Comment:
>up to the custom operation implementor [](start = 76, length = 38)

Different implementations may use different IDs with the same meaning. This means that client code will only be able to run on a single quantum processor or a set of quantum processors with standardized the IDs. Have we evaluated long term implications on Q# ecosystem?

Reply:
Correct. Custom operation functionality requires user to be aware of which custom processor it is running against. it is users responsibility to ensure correct quantum processor. But user is empowered to do this.

---
In reply to: [413476401](https://github.com/microsoft/qsharp-runtime/pull/172#discussion_r413476401) [](ancestors = 413476401)

Reply:
This is a very generic API. Unfortunately, it can fragment the ecosystem and promote writing code that is hard to read and maintain. An analogy in classical programming would be something like void *DoAnything(int id, void*). What other options have we considered to address this scenario?

---
In reply to: [414257458](https://github.com/microsoft/qsharp-runtime/pull/172#discussion_r414257458) [](ancestors = 414257458,413476401)

Reply:
@viktorveis: As detailed above, the existing emulation feature has been used well in the libraries to extend the simulation framework with new accelerated functionality (e.g.: `EstimateFrequencyA`). That works especially well when a pure-Q# fallback can be offered for portability across target machines. I definitely take the point that this extension mechanism could be documented better, or that alternative mechanisms could be provided (e.g.: the attribute-based design discussed above) in order to better enable advanced users looking to extend the simulation runtime, but I think that would be best discussed in the context of a formal feature request issue where the various alternatives can be compared and evaluated against our Q# API Design Principles, and against the need to maintain portability across simulators.